packages/wasm/src/index.ts

```diff
@@ -40,26 +89,66 @@ export function wasm(options: RollupWasmOptions = {}): Plugin {
           }
         }
 
-        if (imports && !sync) {
-          return WebAssembly.instantiate(buf, imports)
-        } else if (!imports && !sync) {
-          return WebAssembly.compile(buf)
-        } else {
+        if(sync) {
           var mod = new WebAssembly.Module(buf)
           return imports ? new WebAssembly.Instance(mod, imports) : mod
+        } else {
+          return _instantiateOrCompile(buf, imports, false)
         }
       }
     `.trim(),
 
     transform(code, id) {
       if (code && /\.wasm$/.test(id)) {
-        const src = Buffer.from(code, 'binary').toString('base64');
-        const sync = syncFiles.indexOf(id) !== -1;
-        return `export default function(imports){return _loadWasmModule(${+sync}, '${src}', imports)}`;
+        const isSync = syncFiles.indexOf(id) !== -1;
+        const filepath = copies[id] ? `'${copies[id]}'` : null;
+        let src;
+
+        if (filepath === null) {
+          assert(!isSync, 'non-inlined files can not be `sync`.');
+          src = Buffer.from(code, 'binary').toString('base64');
+          src = `'${src}'`;
+        } else {
+          src = null;
+        }
+
+        return `export default function(imports){return _loadWasmModule(${+isSync}, ${filepath}, ${src}, imports)}`;
       }
       return null;
+    },
+    generateBundle: async function write(outputOptions) {
+      // can't generate anything if we can't determine the output base
+      if (!outputOptions.dir && !outputOptions.file) {
+        return;
+      }
+
+      const base = outputOptions.dir || path.dirname(outputOptions.file);
+
+      await makeDir(base);
+
+      await Promise.all(
+        Object.keys(copies).map(async (name) => {
+          const output = copies[name];
+          // Create a nested directory if the fileName pattern contains
+          // a directory structure
+          const outputDirectory = path.join(base, path.dirname(output));
+          await makeDir(outputDirectory);
+          return copy(name, path.join(base, output));
```

Comment:
the preferred method for copying files to the bundle destination is similar to this: https://github.com/rollup/plugins/blob/23df786d92318d5e142e60c5e980a7698aa2c617/packages/html/lib/index.js#L104-L113

you're already performing a read on line 31 above, so stashing the buffer there to pass to `this.emitFile` here shouldn't cause any trouble.

Reply:
Ok, I chose this again because of code from the URL plugin:

https://github.com/rollup/plugins/blob/23df786d92318d5e142e60c5e980a7698aa2c617/packages/url/src/index.js#L84-L91

is there any reason for it being used there instead of `this.emitFile`?

Reply:
yeah, a lot of the plugins still have legacy code in them from much earlier versions of rollup. just one of those legacy things that we haven't gotten around to changing. The `fs` methods are still valid, but we prefer `this.emitFile` now.

Reply:
Fixed now, I just named it "Rollup WASM Asset".